working/4271 - static enough metaprogramming/proposal.md

```diff
@@ -0,0 +1,1222 @@
+deployment. Java ecosystem went with the second option: shrinking tools
+more-or-less ignore reflection and developers have to [manually inform
+toolchain][android-shrink-code] about the program elements which are accessed
+reflectively.
+
+> [!NOTE]
+>
+> There has been a number of attempts to statically analyze reflection in Java
+> projects, but they have all hit issues around scalability and precision of the
+> analysis. See:
+>
+> - [Reflection Analysis for Java][paper-java-suif-reflection]
+> - [Understanding and Analyzing Java Reflection][paper-java-reflection-2019]
+> - [Challenges for Static Analysis of Java Reflection – Literature Review and Empirical Study](paper-java-reflection-challenges)
+>
+> Graal VM Native Image (AOT compiler for Java) attempts to fold away as much of
+> reflection uses as it can, but otherwise just like ProGuard and similar tools
+> [relies][java-native-image-reflection] on the developer to inform compiler
+> about reflection uses it could not resolve statically.
+>
+> R8 (Android bytecode shrinker) has a special
+> [troubleshooting][r8-troubleshooting] section in its `README` to cover obscure
+> situations which might arise if developer fails to properly configure ProGuard
+> rules to cover reflection uses.
+>
+> [Reflekt: a Library for Compile-Time Reflection in Kotlin][paper-reflekt]
+> describes a compiler plugin based compile time reflection system similar in
+> some ways to `reflectable`.
+>
+> [Compile-time Reflection and Metaprogramming for Java][paper-miao-siek] covers
+> a metaprograming system which proposes metaprogramming system based on
+> compile-time reflection.
+
+Dart initially went with the first option and tried to make `dart:mirrors` _just
+work_ when compiling Dart to JavaScript. However, rather quickly `dart2js` team
+started facing performance and code size issues caused by `dart:mirrors` in
+large Web applications. So they switched gears and tried the second option:
+introduced [`@MirrorsUsed`][dart-mirrors-used] annotation. However it provided
+only a temporary and partial reprieve from the problems and was eventually
+abandoned together with `dart:mirrors`.
+
+There were two other attempts to address code size issues caused by mirrors,
+while retaining some amount of reflective capabilities: now abandoned package
+[`smoke`][pkg-smoke] and still maintained package
+[`reflectable`][pkg-reflectable]. Both of these apply similar approach: instead
+of relying on the toolchain to provide unrestricted reflection, have developer
+opt-in into specific reflective capabilities for specific parts of the program
+then generate a pile of auxiliary Dart code implementing these capabilities.
+
+> [!NOTE]
+>
+> Another exploration similar in nature was
+> (go/const-tree-shakeable-reflection-objects)[http://go/const-tree-shakeable-reflection-objects].
+
+Fundamentally both of these approaches were dead ends and Web applications
```

Comment:
Fwiw, I wouldn't say that the const reflection proposal was a dead end at all, we just simply didn't end up doing it but there wasn't any sort of blocker or anything that I remember encountering.